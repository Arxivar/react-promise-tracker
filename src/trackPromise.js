import { Emitter } from './tinyEmitter';
import { PROGRESS_UPDATE, DEFAULT_GROUP } from './constants';
import { event } from './utils';

/**
 * Emitter singleton.
 */
export const emitter = new Emitter();

/**
 * Promise registry based on counters and its utility methods.
 */
const countMap = {};
const incrementCount = group => ++countMap[group] || ((countMap[group] = 1), 1);
const decrementCount = group => --countMap[group];
export const inProgress = group => {
  const counter = countMap[group || DEFAULT_GROUP];
  return Number.isInteger(counter) ? counter > 0 : undefined;
}

/**
 * Track a promise by increasing its group counter by one & programm its
 * untracking once the promise has settled.
 */
export const trackPromise = (promise, group) => {
  group = group || DEFAULT_GROUP;
  emitter.emit(event(PROGRESS_UPDATE, group), incrementCount(group) > 0);

  const onResolveHandler = () => untrack(group);
  promise.then(onResolveHandler, onResolveHandler);

  return promise;
};

/**
 * Untrack a promise by decrementing one to its group counter.
 */
const untrack = group => {
  emitter.emit(event(PROGRESS_UPDATE, group), decrementCount(group) > 0);
};

// TODO: Error management. Catch unhandled errors and throw events?
