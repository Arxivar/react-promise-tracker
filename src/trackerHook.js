import React from 'react'
import { emitter, getCounter, promiseCounterUpdateEventId } from './trackPromise';

const usePromiseTracker = () => {
  const [promiseInProgress, setPromiseInProgress]= React.useState(false);
  // pending area

  React.useEffect(()=> {
    emitter.on(promiseCounterUpdateEventId, (anyPromiseInProgress, area) => {
      // pending manage area
      setPromiseInProgress(anyPromiseInProgress);
    });

    return () => {emitter.off(promiseCounterUpdateEventId);
    }
  },[]);

  return {promiseInProgress}
}