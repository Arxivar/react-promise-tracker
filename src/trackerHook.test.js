import React from 'react';
import * as trackPromiseAPI from './trackPromise';
import { usePromiseTracker } from './trackerHook';
import { randomBytes } from 'crypto';

/**
 * Config section.
 */
beforeEach(() => {
  jest.clearAllTimers();
  jest.useFakeTimers(); // Place it before each test.
});

afterEach(() => {
  jest.runAllTimers();
});

afterAll(() => {
  jest.useRealTimers();
});

/**
 * Common stubs.
 */
const testGroup = 'testGroup';
const randomGroup = () => randomBytes(10).toString('hex');
const PROGRESS = 'PROGRESS';
const IDLE = 'IDLE';
const NO_TRACKING = 'noTracking';
const TrackingComponent = ({ group, delay }) => {
  const { progress } = usePromiseTracker({ group, delay });
  return progress === true ? <p>{PROGRESS}</p> :
    progress === false ? <p>{IDLE}</p> : <p>{NO_TRACKING}</p>;
};
const createFakePromise = (duration = 1000) =>
  new Promise(resolve => setTimeout(() => resolve()), duration);

/**
 * Test suites
 */
describe('usePromiseTracker', () => {
  describe('basic testing', () => {
    it('renders without crashing', () => {
      // Arrange & Act
      const component = mount(<TrackingComponent />);

      // Assert
      expect(component).toMatchSnapshot();
    });

    it('renders NO TRACKING when no promise is tracked yet for the target group', () => {
      // Arrange & Act
      const component = mount(<TrackingComponent group={randomGroup()} />);

      // Assert
      expect(component.text()).toEqual(NO_TRACKING);
    });

    it('renders PROGRESS when the tracked promise is pending', () => {
      // Arrange
      const promise = createFakePromise(10);
      trackPromiseAPI.trackPromise(promise, testGroup);

      // Act
      const component = mount(<TrackingComponent group={testGroup} />);

      // Assert
      expect(component.text()).toEqual(PROGRESS);
    });

    it('renders IDLE once the tracked promise has settled', done => {
      // Arrange
      const promise = createFakePromise(10);
      trackPromiseAPI.trackPromise(promise, testGroup);

      // Act
      const component = mount(<TrackingComponent group={testGroup} />);
      jest.runAllTimers();

      // Assert
      promise.then(() => {
        expect(component.text()).toEqual(IDLE);
        done();
      })
    });
  });

  describe('several promises tracked', () => {
    it('renders PROGRESS -> IDLE with multiple overlapping promises, tracked before mount', done => {
      // Arrange
      const group = randomGroup();
      const promiseA = createFakePromise(10);
      const promiseB = createFakePromise(20);

      // Act & Assert
      trackPromiseAPI.trackPromise(promiseA, group);
      trackPromiseAPI.trackPromise(promiseB, group);

      // 1. Mount. Expect PROGRESS.
      const component = mount(<TrackingComponent group={group} />);
      expect(component.text()).toEqual(PROGRESS);

      // 2. First promise settled, second one still pending. Expect PROGRESS.
      jest.advanceTimersByTime(15);
      expect(component.text()).toEqual(PROGRESS);

      // 3. Both promises settled. Expect IDLE.
      jest.runAllTimers();
      Promise.all([promiseA, promiseB]).then(() => {
        expect(component.text()).toEqual(IDLE);
        done();
      })
    });

    it('renders NO_TRACKING -> PROGRESS -> IDLE with multiple overlapping promises, tracked after mount', done => {
      // Arrange
      const group = randomGroup();
      const promiseA = createFakePromise(10);
      const promiseB = createFakePromise(20);

      // Act & Assert
      const component = mount(<TrackingComponent group={group} />);

      // 1. On mount, no promises tracked. Expect NO_TRACKING.
      expect(component.text()).toEqual(NO_TRACKING);

      // 2. Now track promises. Expect PROGRESS.
      trackPromiseAPI.trackPromise(promiseA, group);
      trackPromiseAPI.trackPromise(promiseB, group);
      expect(component.text()).toEqual(PROGRESS);

      // 3. First promise settled, second one still pending. Expect PROGRESS.
      jest.advanceTimersByTime(15);
      expect(component.text()).toEqual(PROGRESS);

      // 4. Both promises settled. Expect IDLE.
      jest.runAllTimers();
      Promise.all([promiseA, promiseB]).then(() => {
        expect(component.text()).toEqual(IDLE);
        done();
      })
    });

    it('renders NO_TRACKING -> PROGRESS -> IDLE -> PROGRESS -> IDLE with multiple non-overlapping promises', done => {
      // Arrange
      const group = randomGroup();
      const promiseA = createFakePromise(10);

      // Act & Assert
      const component = mount(<TrackingComponent group={group} />);

      // 1. On mount, no promises tracked. Expect NO_TRACKING.
      expect(component.text()).toEqual(NO_TRACKING);

      // 2. Track promise A. Expect PROGRESS.
      trackPromiseAPI.trackPromise(promiseA, group);
      expect(component.text()).toEqual(PROGRESS);

      // 3. Let promise A resolve. Expect IDLE.
      jest.runOnlyPendingTimers();
      promiseA.then(() => {
        expect(component.text()).toEqual(IDLE);

        // 4. Track promise B. Expect PROGRESS.
        const promiseB = createFakePromise(10);
        trackPromiseAPI.trackPromise(promiseB, group);
        expect(component.text()).toEqual(PROGRESS);

        // 5. Let promise B resolve. Expect IDLE.
        jest.runAllTimers();
        promiseB.then(() => {
          expect(component.text()).toEqual(IDLE);
          done();
        });
      });
    });
  });

  describe('delay timing', () => {
    it('renders NO_TRACKING -> PROGRESS -> IDLE with active delay, promise tracked before mount', done => {
      // Arrange
      const group = randomGroup();
      const promise = createFakePromise(50);

      // Act & Assert

      // 1. Track promise and Mount. Expect NO_TRACKING.
      trackPromiseAPI.trackPromise(promise, group);
      const component = mount(<TrackingComponent group={group} delay={20} />);
      expect(component.text()).toEqual(NO_TRACKING);

      // 2. Delay window completed. Expect PROGRESS.
      jest.advanceTimersByTime(25);
      expect(component.text()).toEqual(PROGRESS);

      // 3. Settled promise. Expect IDLE.
      jest.runAllTimers();
      promise.then(() => {
        expect(component.text()).toEqual(IDLE);
        done();
      });
    });

    it('renders NO_TRACKING -> PROGRESS -> IDLE with active delay, promise tracked after mount', done => {
      // Arrange
      const group = randomGroup();
      const promise = createFakePromise(50);

      // Act & Assert

      // 1. Mount. Expect NO_TRACKING.
      const component = mount(<TrackingComponent group={group} delay={20} />);
      expect(component.text()).toEqual(NO_TRACKING);

      // 2. Track promise. Expect NO_TRACKING.
      trackPromiseAPI.trackPromise(promise, group);
      expect(component.text()).toEqual(NO_TRACKING);

      // 3. Delay window completed. Expect PROGRESS.
      jest.advanceTimersByTime(25);
      expect(component.text()).toEqual(PROGRESS);

      // 4. Settled promise. Expect IDLE.
      jest.runAllTimers();
      promise.then(() => {
        expect(component.text()).toEqual(IDLE);
        done();
      });
    });
  });
});
