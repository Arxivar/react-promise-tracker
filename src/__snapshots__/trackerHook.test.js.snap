// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usePromiseTracker basic testing renders without crashing 1`] = `
<TrackingComponent>
  <p>
    noTracking
  </p>
</TrackingComponent>
`;
